proof/refine/ARM/Ipc_R.thy

```diff
@@ -3560,160 +3560,622 @@ lemma cteInsert_ct'[wp]:
+         setQueue_cur untyped_ranges_zero_lift threadSet_ifunsafe'T threadSet_cur crunch_wps)
+
+lemma schedContextDonate_valid_pspace':
+  "\<lbrace>valid_pspace' and tcb_at' tcbPtr\<rbrace> schedContextDonate scPtr tcbPtr \<lbrace>\<lambda>_. valid_pspace'\<rbrace>"
+  by (wpsimp wp: schedContextDonate_valid_objs' simp: valid_pspace'_def)
+
+lemma schedContextDonate_sch_act_wf [wp]:
+  "schedContextDonate scPtr tcbPtr \<lbrace>\<lambda>s. sch_act_wf (ksSchedulerAction s) s\<rbrace>"
+  apply (simp only: schedContextDonate_def)
+  apply (wpsimp wp: threadSet_sch_act threadSet_wp)
+       apply (rule_tac Q="\<lambda>_ s. sch_act_wf (ksSchedulerAction s) s" in hoare_strengthen_post[rotated])
+        apply (fastforce simp: sch_act_wf_def ct_in_state'_def tcb_in_cur_domain'_def
+                               pred_tcb_at'_def obj_at'_def projectKO_eq projectKO_tcb
+                        split: if_splits)
+       apply wpsimp+
+  done
+
+lemma schedContextDonate_if_live_then_nonz_cap':
+  "\<lbrace>\<lambda>s. if_live_then_nonz_cap' s \<and> valid_objs' s \<and>
+        ex_nonz_cap_to' tcbPtr s \<and> ex_nonz_cap_to' scPtr s\<rbrace>
+   schedContextDonate scPtr tcbPtr
+   \<lbrace>\<lambda>_. if_live_then_nonz_cap'\<rbrace>"
+  unfolding schedContextDonate_def
+  by (wpsimp wp: threadSet_iflive'T setSchedContext_iflive' hoare_vcg_all_lift threadSet_cap_to'
+           simp: conj_ac cong: conj_cong | wp hoare_drop_imps | fastforce simp: tcb_cte_cases_def)+
+
+lemma schedContextDonate_valid_idle':
+  "\<lbrace>\<lambda>s. valid_idle' s \<and> tcbPtr \<noteq> ksIdleThread s \<and>
+        obj_at' (\<lambda>x. scTCB x \<noteq> Some (ksIdleThread s)) scPtr s\<rbrace>
+   schedContextDonate scPtr tcbPtr
+   \<lbrace>\<lambda>_. valid_idle'\<rbrace>"
+  apply (simp only: schedContextDonate_def)
+  apply (wp threadSet_idle' setSchedContext_valid_idle')
+       apply (rule_tac Q="\<lambda>_ s. tcbPtr \<noteq> ksIdleThread s" in hoare_strengthen_post; wpsimp)
+      apply (rule_tac Q="\<lambda>_ s. valid_idle' s \<and> scPtr \<noteq> idle_sc_ptr \<and> tcbPtr \<noteq> ksIdleThread s"
+                   in hoare_strengthen_post; wpsimp)
+        apply (wpsimp wp: threadSet_idle' hoare_drop_imps threadSet_idle')
+       apply (rule_tac Q="\<lambda>_ s. valid_idle' s \<and> scPtr \<noteq> idle_sc_ptr \<and>
+                                tcbPtr \<noteq> ksIdleThread s \<and> from \<noteq> ksIdleThread s"
+                    in hoare_strengthen_post)
+        apply wpsimp+
+  apply (auto simp: obj_at'_def projectKO_eq projectKO_sc valid_idle'_def)
+  done
+
+lemma schedContextDonate_valid_release_queue'[wp]:
+  "schedContextDonate scPtr tcbPtr \<lbrace>valid_release_queue'\<rbrace>"
+  unfolding schedContextDonate_def
+  by (wpsimp wp: threadSet_valid_release_queue' simp: obj_at_ko_at'_eq valid_release_queue'_def
+      | rule hoare_strengthen_post[where Q="\<lambda>_. valid_release_queue'"])+
+
+lemma schedContextDonate_assume_asserts:
```

Comment:
A little rule to introduce the `sc_at'` fact is nice, but I'd prefer to not include the `sym_refs` assert here; we may add more asserts in the future, which would mean updating this too, and it's easy enough to use `stateAssert_sp` to add the assert to the preconditions

Reply:
> A little rule to introduce the `sc_at'` fact is nice, but I'd prefer to not include the `sym_refs` assert here; we may add more asserts in the future, which would mean updating this too, and it's easy enough to use `stateAssert_sp` to add the assert to the preconditions

Actually, I am not sure what exactly mean, Michael. Which assert do you prefer not including?

Reply:
I'd rather not have `sym_refs (state_refs_of' s)` in the precondition here. I think it would be better to have a rule to add the `sc_at'` bit to the preconditions, and just use `stateAssert_sp` separately to get the `sym_refs` bits. That way, if we add further asserts to `schedContextDonate` later on, we won't need to update this proof too. Not that this is a big deal

Reply:
Right. I see. We might want to look again why Ryan wants `sym_refs` in the later proofs.

Reply:
> I'd rather not have `sym_refs (state_refs_of' s)` in the precondition here. I think it would be better to have a rule to add the `sc_at'` bit to the preconditions, and just use `stateAssert_sp` separately to get the `sym_refs` bits. That way, if we add further asserts to `schedContextDonate` later on, we won't need to update this proof too. Not that this is a big deal
It's not in the precondition, this is just a rule to add it to the precondition. To make it clearer, I've removed this lemma and added the proofs to `schedContextDonate_invs'` instead.